dev/upgrade.sh

```diff
@@ -0,0 +1,24 @@
+# https://github.com/jazzband/pip-tools/issues/625
+
+PYTHON_VERSION=$(python -c 'import sys; print(".".join(map(str, sys.version_info[:2])))')
+
+if [[ -z "$PYTHON_VERSION" ]]; then
+    echo -e "\e[31;1mERROR: environment variable PYTHON_VERSION needs to be defined!\e[0m"
+    exit 1
+fi
+
+mkdir -p dev/$PYTHON_VERSION &&
+    python dev/extract_install_requires.py &&
+    cp dev/requirements*.in dev/$PYTHON_VERSION/ &&
+    rm dev/$PYTHON_VERSION/requirements-dev.in &&
+    pip-compile --upgrade \
+        dev/requirements*.in \
+        -o dev/$PYTHON_VERSION/requirements-dev.txt &&
+    for in_file in $(ls dev/$PYTHON_VERSION/requirements*.in); do
+        echo -e "-c requirements-dev.txt\n$(cat ${in_file})" >${in_file}
+        pip-compile "${in_file}"
```

Comment:
You could try using a pipe `|` here. So something like
```
echo -e "-c requirements-dev.txt\n$(cat ${in_file})" | pip-compile
```

Reply:
Yeah true, but I thought it (theoretically) more transparent to write this line to the output `.in` file, even if it is not committed.

Reply:
To me using the stdin/out with pipes is much more elegant. What confused me is that this `in_file` is overwritten. So was not that transparent to me :grimacing:

Reply:
This is the way to prepend to a file in bash. The line has to be prepended (not `>>`), because the`pip` constraints have to be specified first.

Reply:
> To me using the stdin/out with pipes is much more elegant.

Note that the files have to be written to the `3.6`, `3.7`, `3.8` subfolders, because there are relative path references to the `.txt` files. These files are in the subfolders, not the main `reqs` folder.

Reply:
I know, well you are "prepending" by overwriting the file completely. I this actually what we want? That in file should stay untouched right?
Anyways the pipe would be much more transparent to me

Reply:
The `in_file` is a copy.
https://github.com/ComPWA/expertsystem/blob/925f344512745b03a055531268784762aecbdb4e/reqs/upgrade.sh#L12

I just tried with a pipe, but the problem remains that `pip-compile` needs to know which `requirements-dev.txt` file to use as constraints. For that it needs to know which subfolder to use, and since `pip-compile` does not support the `-c` constraint flag, this needs to be done in an `.in` file _that is placed in that subfolder_ (so that it can refer to the `requirements-dev.txt` that has been dumped to that folder previously).

Reply:
Ok maybe add a comment that `pip-compile` does not support this, so that no one else stumbles there